include/dd/NoiseFunctionality.hpp

```diff
@@ -0,0 +1,714 @@
+/*
+* This file is part of the MQT DD Package which is released under the MIT license.
+* See file README.md or go to https://www.cda.cit.tum.de/research/quantum_dd/ for more information.
+*/
+
+#ifndef DDSIM_NOISEFUNCTIONALITY_HPP
+#define DDSIM_NOISEFUNCTIONALITY_HPP
+
+#include "dd/Definitions.hpp"
+#include "dd/Export.hpp"
+
+#include <random>
+#include <utility>
+
+using CN = dd::ComplexNumbers;
+
+namespace dd {
+
+    // noise operations available for deterministic noise aware quantum circuit simulation
+    enum NoiseOperations : std::uint8_t {
+        amplitudeDamping,
+        phaseFlip,
+        depolarization,
+        identity
+    };
+
+    template<class DDPackage>
+    class StochasticNoiseFunctionality {
+    public:
+        StochasticNoiseFunctionality(const std::unique_ptr<DDPackage>& package,
+                                     dd::QubitCount                    nQubits,
+                                     double                            gateNoiseProbability,
+                                     double                            amplitudeDampingProb,
+                                     double                            multiQubitGateFactor,
+                                     std::vector<dd::NoiseOperations>  noiseEffects):
+            package(package),
+            nQubits(nQubits),
+            dist(0.0, 1.0L),
+            noiseProbability(gateNoiseProbability),
+            noiseProbabilityMulti(gateNoiseProbability * multiQubitGateFactor),
+            sqrtAmplitudeDampingProbability({std::sqrt(amplitudeDampingProb), 0}),
+            oneMinusSqrtAmplitudeDampingProbability({std::sqrt(1 - amplitudeDampingProb), 0}),
+            sqrtAmplitudeDampingProbabilityMulti({std::sqrt(gateNoiseProbability) * multiQubitGateFactor, 0}),
+            oneMinusSqrtAmplitudeDampingProbabilityMulti({std::sqrt(1 - multiQubitGateFactor * amplitudeDampingProb), 0}),
+            noiseEffects(std::move(noiseEffects)) {
+            ampDampingFalse      = dd::GateMatrix({dd::complex_one, dd::complex_zero, dd::complex_zero, oneMinusSqrtAmplitudeDampingProbability});
+            ampDampingFalseMulti = dd::GateMatrix({dd::complex_one, dd::complex_zero, dd::complex_zero, oneMinusSqrtAmplitudeDampingProbabilityMulti});
+            ampDampingTrue       = dd::GateMatrix({dd::complex_zero, sqrtAmplitudeDampingProbability, dd::complex_zero, dd::complex_zero});
+            ampDampingTrueMulti  = dd::GateMatrix({dd::complex_zero, sqrtAmplitudeDampingProbabilityMulti, dd::complex_zero, dd::complex_zero});
+            identityDD           = package->makeIdent(nQubits);
+            package->incRef(identityDD);
+        }
+
+        ~StochasticNoiseFunctionality() {
+            package->decRef(identityDD);
+        }
+
+    protected:
+        const std::unique_ptr<DDPackage>&      package;
```

Comment:
I am really surprised that this `const` here works since you are conducting operations that alter the DD Package (such as the `incRef` call in the constructor.

Reply:
To me it seems like the pointer is constant and not the object the pointer points to.

Reply:
True, still surprised.